ffconv/stream_processors.py

```diff
@@ -0,0 +1,240 @@
+        self.target_level = profile[self.media_type]['level']
+        self.target_preset = profile[self.media_type]['preset']
+        self.target_quality = profile[self.media_type]['quality']
+
+    @property
+    def must_convert(self):
+        """
+        Conversion check: besides base check (codec compatibility), check that
+        reference frames are acceptable.
+        """
+        return any((super(VideoProcessor, self).must_convert,
+                    self.refs > self.max_refs))
+
+    def clean_up(self):
+        """
+        Cleanup is no-op for video, because the stream is not modified.
+        """
+        pass
+
+    def convert(self):
+        """
+        Convert the video stream according to the selected target values,
+        creating a temporary video file to use by merger.
+        """
+        cmd = ['ffmpeg', '-i', self.input, '-map', '0:{}'.format(self.index),
+               '-c:v', self.target_codec, '-preset', str(self.target_preset),
+               '-crf', str(self.target_quality), '-profile:v',
+               self.target_profile, '-level', self.target_level, self.output]
+        execute_cmd(cmd)
+
+
+class AudioProcessor(StreamProcessor):
+    """
+    Audio stream processor, provides specific audio conversion functionality.
+    """
+    media_type = 'audio'
+
+    def _init_stream(self, input, stream, profile):
+        """
+        Set audio-specific input and target specs,
+        """
+        # Set number of channels in input stream
+        self.channels = int(stream['channels'])
+
+        # Set target quality and channels
+        self.target_quality = profile[self.media_type]['quality']
+        self.target_channels = int(profile[self.media_type]['channels'])
+
+    @property
+    def must_convert(self):
+        """
+        Conversion check: besides base check (codec compatibility), check that
+        number of channels is acceptable.
+        """
+        return any((super(AudioProcessor, self).must_convert,
+                    self.channels != self.target_channels))
+
+    def clean_up(self):
+        """
+        Cleanup is no-op for audio, because the stream is not modified.
+        """
+        pass
+
+    def convert(self):
+        """
+        Convert the audio stream according to the selected target values,
+        creating a temporary audio file to use by merger.
+        """
+        cmd = ['ffmpeg', '-i', self.input, '-map', '0:{}'.format(self.index),
+               '-c:a', self.target_codec, '-q:a', str(self.target_quality),
+               '-ac:0', str(self.target_channels), self.output]
+        execute_cmd(cmd)
+
+
+class SubtitleProcessor(StreamProcessor):
+    """
+    Subtitle stream processor, provides specific subtitle conversion
+    functionality.
+    """
+    media_type = 'subtitle'
+
+    # Override super property to always convert (we always want to clean-up)
+    must_convert = True
+
+    def _init_stream(self, input, stream, profile):
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Redefining built-in 'input' [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://localhost:9000/coding_rules#rule_key=Pylint%3AW0622)